src/snapshot.rs

```diff
@@ -0,0 +1,302 @@
+        // Acquire a lock on the snapshot vector to access the desired snapshot
+        let mut snapshots = self
+            .tree
+            .snapshots
+            .write()
+            .map_err(|_| TrieError::SnapshotMutexError)?;
+
+        // Try to get the mutable reference to the snapshot, returning `SnapshotNotFound` if not found
+        let snapshot = snapshots
+            .get_mut(&self.id)
+            .ok_or(TrieError::SnapshotNotFound)?;
+
+        // Insert the key-value pair into the snapshot
+        snapshot.insert(key, value)?;
+
+        Ok(())
+    }
+
+    pub fn get<K: Key>(&self, key: &K) -> Result<Option<(V, u64)>, TrieError> {
+        // Acquire a lock on the snapshot vector to access the desired snapshot
+        let snapshots = self
+            .tree
+            .snapshots
+            .read()
+            .map_err(|_| TrieError::SnapshotMutexError)?;
+
+        // Try to get the snapshot at the specified id, returning `SnapshotNotFound` if not found
+        let snapshot = snapshots.get(&self.id).ok_or(TrieError::SnapshotNotFound)?;
+
+        Ok(snapshot.get(key))
+    }
+
+    /// Closes the snapshot referred to by this SnapshotPointer, releasing resources associated with it.
+    pub fn close(&mut self) -> Result<(), TrieError> {
+        // Call the close method of the Tree with the id of the snapshot to close it
+        self.tree.close(self.id)?;
+        Ok(())
+    }
+}
+
+/// Represents a snapshot of the data within the Trie.
+pub struct Snapshot<P: PrefixTrait, V: Clone> {
+    pub(crate) id: u64,
+    pub(crate) ts: u64,
+    pub(crate) root: RwLock<Arc<Node<P, V>>>,
+    pub(crate) readers: RwLock<HashMap<u64, ()>>,
+    pub(crate) max_active_readers: AtomicU64,
+    pub(crate) closed: bool,
+}
+
+impl<P: PrefixTrait, V: Clone> Snapshot<P, V> {
+    /// Creates a new Snapshot instance with the provided snapshot_id and root node.
+    pub fn new_snapshot(snapshot_id: u64, root: Arc<Node<P, V>>) -> Snapshot<P, V> {
+        Snapshot {
+            id: snapshot_id,
+            ts: root.ts() + 1,
+            root: RwLock::new(root),
+            readers: RwLock::new(HashMap::new()),
+            max_active_readers: AtomicU64::new(0),
+            closed: false,
+        }
+    }
+
+    pub fn new_reader(&self) -> Result<IterationPointer<P, V>, TrieError> {
+        let reader_id = self.max_active_readers.fetch_add(1, Ordering::SeqCst);
+        let mut readers = self.readers.write().map_err(|_| TrieError::MutexError)?;
+        readers.insert(reader_id, ());
+        let root = self.root.read().unwrap();
+        Ok(IterationPointer::new(self, root.clone(), reader_id))
+    }
+
+    pub fn close_reader(&self, reader_id: u64) -> Result<(), TrieError> {
+        let mut readers = self.readers.write().map_err(|_| TrieError::MutexError)?;
+        readers.remove(&reader_id);
+        self.max_active_readers.fetch_sub(1, Ordering::SeqCst);
+        Ok(())
```

Comment:
That is an interesting point! I'll need to think about it because snapshots are opened per transaction, so I'd think the action taker on the snapshot would have the responsibility to close it